Marlin/src/HAL/STM32F1/HAL_MinSerial.cpp

```diff
@@ -0,0 +1,91 @@
+/**
+ * Marlin 3D Printer Firmware
+ * Copyright (c) 2020 MarlinFirmware [https://github.com/MarlinFirmware/Marlin]
+ *
+ * Based on Sprinter and grbl.
+ * Copyright (c) 2011 Camiel Gubbels / Erik van der Zalm
+ * Copyright (c) 2017 Victor Perez
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ */
+#ifdef __STM32F1__
+
+#include "../../inc/MarlinConfig.h"
+
+#if ENABLED(POST_MORTEM_DEBUGGING)
+
+#include "../shared/HAL_MinSerial.h"
+#include "watchdog.h"
+
+#include <libmaple/usart.h>
+#include <libmaple/rcc.h>
+#include <libmaple/nvic.h>
+
+/* Instruction Synchronization Barrier */
+#define isb() __asm__ __volatile__ ("isb" : : : "memory")
+
+/* Data Synchronization Barrier */
+#define dsb() __asm__ __volatile__ ("dsb" : : : "memory")
+
+static void TXBegin() {
+  // We use MYSERIAL0 here, so we need to figure out how to get the linked register
+  struct usart_dev* dev = MYSERIAL0.c_dev();
+
+  // Or use this if removing libmaple
+  // int irq = dev->irq_num;
+  // int nvicUART[] = { NVIC_USART1 /* = 37 */, NVIC_USART2 /* = 38 */, NVIC_USART3 /* = 39 */, NVIC_UART4 /* = 52 */, NVIC_UART5 /* = 53 */ };
+  // Disabling irq means setting the bit in the NVIC ICER register located at
+  // Disable UART interrupt in NVIC
+  nvic_irq_disable(dev->irq_num);
+
+  // Use this if removing libmaple
+  //NVIC_BASE->ICER[1] |= (1<< irq - 32);
+
+  // We NEED memory barriers to ensure Interrupts are actually disabled!
+  // ( https://dzone.com/articles/nvic-disabling-interrupts-on-arm-cortex-m-and-the )
+  dsb();
+  isb();
+
+  rcc_clk_disable(dev->clk_id);
+  rcc_clk_enable(dev->clk_id);
+
+  usart_reg_map *regs = dev->regs;
+  regs->CR1 = 0; // Reset the USART
+  regs->CR2 = 0; // 1 stop bit
+
+  // If we don't touch the BRR (baudrate register), we don't need to recompute. Else we would need to call
+  usart_set_baud_rate(dev, 0, BAUDRATE);
+
+  regs->CR1 = (USART_CR1_TE | USART_CR1_UE); // 8 bits, no parity, 1 stop bit
+}
+
+// A SW memory barrier, to ensure GCC does not overoptimize loops
+#define sw_barrier() __asm__ volatile("": : :"memory");
+static void TX(char c) {
+  struct usart_dev* dev = MYSERIAL0.c_dev();
```

Comment:
I don't think this should assume `MYSERIAL0` is the correct port for debugging.

Many 32-bit boards use native USB for serial communication. This will need to always use a hardware UART.

I think the serial port to use should be able to be specified _somehow_. I do not think it should have to be one of the "normal" Marlin `SERIAL_PORT`s.

Reply:
It's not clear to me. If I understand the serial port selection logic correctly (I'm not sure), `MYSERIAL0` would map to the `Configuration.h` *first*  `SERIAL_PORT`. 

For using a USB port, I don't know if it's possible with interrupt disabled (... ahem...). If it is, then a new TX function must be written and then, we just need to set the function pointer on this new function instead.

Reply:
Ideally, the current serial output port should be used (the one the user is more likely to observe)
I need more guidance from maintainers here. I see two options here:
1. Add a `POST_MORTEM_SERIAL` macro in ConfAdv (by default falling back to MYSERIAL0). And let the user override
2. Find a way to use the current **main** serial port. This means being able to use USB's CDC serial port with interrupt disabled (via polling). I'm currently documenting one how to do that. I won't be able to test however since I don't have any board with this configuration.

Reply:
The appropriate serial port to use is the one which is expected to be connected to a PC-based host (or OctoPrint). A serial port attached to small LCD touch controllers can be ignored. Where this is ambiguous in the configuration, the user should be able to override the first choice. I would put the USB port ahead of all others for this debugging option, or even make it so that it requires the USB port to be chosen when there is one.